Update internationalization documentation (#8065)
flutter/website: src/development/accessibility-and-localization/internationalization.md

  <h4 class="no_toc">What you’ll learn</h4>

  * How to track the device's locale (the user's preferred language).

>   * How to enable locale-specific Material/Cupertino widgets.

Reviewer: ```suggestion
  * How to enable locale-specific Material or Cupertino widgets.
```

```json
"nWombats": "{count, plural, =0{no wombats} =1{1 wombat} other{{ '{{' }}count} wombats}}",
"@nWombats": {
  "description": "A plural message",
  "placeholders": {
    "count": {
      "type": "num"
    }
  }
},
```

Using a plural method is easy enough, just pass it the item count parameter:
<?code-excerpt "gen_l10n_example/lib/main.dart (Placeholder)" remove="/'He|hello|ombat/" replace="/\[/[\n    .../g"?>
```dart
// Examples of internationalized strings.
return Column(
  children: <Widget>[
    ...
    // Returns 'he'
    Text(AppLocalizations.of(context).pronoun('male')),
    // Returns 'she'
    Text(AppLocalizations.of(context).pronoun('female')),
    // Returns 'they'
    Text(AppLocalizations.of(context).pronoun('other')),
  ],
),
```

Similar to plurals, there is a syntax to choose a value based on a `String` placeholder.
This is most often used to support gendered languages. The syntax is
<?code-excerpt "gen_l10n_example/lib/examples.dart (SelectSyntax)"?>
```json
"{selectPlaceholder, select, case{message} ... other{messageOther}}"
```

For example to define a message with a select:
<?code-excerpt "gen_l10n_example/lib/l10n/app_en.arb (Selection)"?>
```json

> "pronoun": "{gender, select, male{he} female{she} other{they}}",

Author: Yes, it's case sensitive. I wrote up something after the code snippet, could you see if it sounds good?

```json
"pronoun": "{gender, select, male{he} female{she} other{they}}",
"@pronoun": {
  "description": "A gendered message",
  "placeholders": {
    "gender": {
      "type": "String"
    }
  }
}
```
Similar to above, just pass in the gender as a parameter:
<?code-excerpt "gen_l10n_example/lib/main.dart (Placeholder)" remove="/'He|hello|ombat/" replace="/\[/[\n    .../g"?>
```dart
// Examples of internationalized strings.
return Column(
  children: <Widget>[
    ...
    // Returns 'he'
    Text(AppLocalizations.of(context)!.pronoun('male')),
    // Returns 'she'
    Text(AppLocalizations.of(context)!.pronoun('female')),
    // Returns 'they'
    Text(AppLocalizations.of(context)!.pronoun('other')),
  ],
);
```

### Messages with numbers and currencies
Numbers and numbers that represent currency values are
displayed very differently in different locales. 
The localizations generation tool in `flutter_localizations` makes use of the
[`intl` package's `NumberFormat` class](https://api.flutter.dev/flutter/intl/NumberFormat-class.html)
to properly format numbers based on the locale and the
desired format.

> The “format” for placeholders whose type is `int`, `double`, or `number`
> can be any one of the following `NumberFormat` constructors.

Reviewer: ```suggestion
The `int`, `double`, and `number` types can use any of the
following `NumberFormat` constructors:
```

<div class="table-wrapper" markdown="1">
| Message "format" value      | Output for 1200000 |
| --------------------------- | ------------------ |
| `compact`                   | "1.2M"             |
| `compactCurrency`*          | "$1.2M"            |
| `compactSimpleCurrency`*    | "$1.2M"            |
| `compactLong`               | "1.2 million"      |
| `currency`*                 | "USD1,200,000.00"  |
| `decimalPattern`            | "1,200,000"        |
| `decimalPercentPattern`*    | "120,000,000%"     |
| `percentPattern`            | "120,000,000%"     |
| `scientificPattern`         | "1E6"              |
| `simpleCurrency`*           | "$1,200,000"       |
{:.table.table-striped}
</div>

> The five starred `NumberFormat` constructors have optional, named parameters. Those
> parameters can be specified as the value of the placeholder’s “optionalParameters” object.
> For example, to specify the optional decimalDigits parameter for "compactCurrency":

Reviewer: ```suggestion
The starred `NumberFormat` constructors in the table offer optional, named parameters.
Those parameters can be specified as the value of the placeholder’s `optionalParameters` object.
For example, to specify the optional `decimalDigits` parameter for `compactCurrency`,
make the following changes to the `lib/l10n/app_en.arg` file:
```

<?code-excerpt "gen_l10n_example/lib/l10n/app_en.arb" skip="34" take="13" replace="/{{/{{ '{{' }}/g;/},$/}/g"?>
```json
"numberOfDataPoints": "Number of data points: {value}",
"@numberOfDataPoints": {
  "description": "A message with a formatted int parameter",
  "placeholders": {
    "value": {
      "type": "int",
      "format": "compactCurrency",
      "optionalParameters": {
        "decimalDigits": 2
      }
    }
  }
}
```

### Messages with dates
Dates strings are formatted in many different ways depending both the locale and the app’s needs.  

Placeholder values with type `DateTime` are formatted with

> [`intl`'s `DateFormat` class](https://api.flutter.dev/flutter/intl/DateFormat-class.html).

Reviewer: ```suggestion
[`intl`'s `DateFormat` class]({{site.api}}/flutter/intl/DateFormat-class.html).
```

There are 41 format variations, identified by the names of their `DateFormat` factory constructors.
In the following example, the `DateTime` value that appears in the `helloWorldOn` message is formatted with `DateFormat.yMd`:

```json
"helloWorldOn": "Hello World on {date}",
"@helloWorldOn": {
  "description": "A message with a date parameter",
  "placeholders": {
    "date": {
      "type": "DateTime",
      "format": "yMd"
    }
  }
}
```

> In an app, when the locale is US English, the following expression would produce  “7/10/1996”. If the locale was Russian, then it would produce “10.07.1996”.

Reviewer: ```suggestion
In an app where the locale is US English, the following expression would produce  “7/10/1996”. In a Russian locale, it would produce “10.07.1996”.
```

a parameter of type `String`; the `hello` method returns a string.

Implement this by replacing the code passed into `Builder`
with the following:

<?code-excerpt "gen_l10n_example/lib/main.dart (Placeholder)" remove="/wombat|Wombats|he'|they|pronoun/"?>
```dart
// Examples of internationalized strings.
return Column(
  children: <Widget>[
    // Returns 'Hello John'
    Text(AppLocalizations.of(context)!.hello('John')),
  ],
);
```

You can also use numerical placeholders to specify multiple values.
Different languages have different ways to pluralize words.
The syntax also supports specifying _how_ a word should be pluralized.
A _pluralized_ message must include a `num` parameter indicating
how to pluralize the word in different situations.  English, for example,
pluralizes "person" to "people", but that doesn't go far enough. The
`message0` plural might be "no people" or "zero people".
The `messageFew` plural might be
"several people", "some people", or "a few people". 
The `messageMany` plural might
be  "most people" or "many people", or "a crowd". 
Only the more general `messageOther` field is required.
The following example shows what options are available:

> ```json

Reviewer: ```suggestion
<!--skip-->
```json
```

> "{countPlaceholder, plural, =0{message0} =1{message1} =2{message2} few{messageFew} many{messageMany} other{messageOther}}"

Reviewer: @johnpryan @khanhnwin @thkim1011  This line is causing our link checker to choke. Apparently. I'd like to merge this PR ASAP, before we freeze the site. Will adding a "skip" message before this JSON code block be enough to fix it?

Author: From the error message, the issue seems to be caused a couple more lines down by this line:
```
<?code-excerpt "gen_l10n_example/lib/l10n/app_en.arb" skip="15" take="10" replace="/{{/{{ '{{' }}/g;/},$/}/g"?>
```
which is strange since this wasn't causing an issue as of Friday.

Author: I added this replacement in the code-excerpt because Jekyll reserves the pattern `{{ }}` as its own special syntax, and the only way to use the characters `{{` in the documentation is to replace it with `{{ '{{' }}`.

);
```

### Configuring the l10n.yaml file

> The `l10n.yaml` file allows you to configure the `gen-l10n` tool,
> such as where all the input files are located,
> where all the output files should be created, and what Dart class
> name to give your localizations delegate.
> The full list of options is described in the table below:

Reviewer: ```suggestion

The `l10n.yaml` file allows you to configure the `gen-l10n` tool
to specify:

* where all the input files are located
* where all the output files should be created
* what Dart class name to give your localizations delegate

For a full list of options, check out the following table:
```
